### PR TITLE
Simplify annotations in `optuna.visualization`

### PR DESCRIPTION
**Pull request by harupy**
## Motivation

https://github.com/optuna/optuna/issues/4508

### CODE REVIEW

File: optuna/visualization/_contour.py [hunk 1]
```diff
@@ -34,31 +31,31 @@
...
     target_name: str
 
 
 def plot_contour(
     study: Study,
-    params: Optional[List[str]] = None,
+    params: list[str] | None = None,
     *,
-    target: Optional[Callable[[FrozenTrial], float]] = None,
+    target: Callable[[FrozenTrial], float] | None = None,
```
Comment (c-bata) [inline optuna/visualization/_contour.py:58]: [question] Though `typing.Optional` seems not to be deprecated, should we replace `typing.Optional` with `| None`? Or is it just a matter of preference?

**Review by c-bata — APPROVED** [review]
Thank you for your pull request. LGTM. I have a question though.

**Comment by c-bata** [comment]
@Alnusjaponica Could you review this PR?

File: optuna/visualization/_pareto_front.py [hunk 2]
```diff
@@ -1,13 +1,11 @@
...
 import json
 from typing import Any
 from typing import Callable
-from typing import Dict
-from typing import List
 from typing import NamedTuple
-from typing import Optional
 from typing import Sequence
-from typing import Tuple
 from typing import Union
```
Comment (Alnusjaponica) [inline optuna/visualization/_pareto_front.py:9]:
```suggestion
```
Is it possible to remove this line and `Union` in the 385th line?

File: optuna/visualization/_parallel_coordinate.py [hunk 3]
```diff
@@ -235,9 +230,9 @@ def _target(t: FrozenTrial) -> float:
                 ticktext=["{:.3g}".format(math.pow(10, x)) for x in tickvals],
             )
         elif _is_categorical(trials, p_name):
-            vocab: DefaultDict[Union[int, str], int] = defaultdict(lambda: len(vocab))
+            vocab: DefaultDict[int | str, int] = defaultdict(lambda: len(vocab))
```
Comment (Alnusjaponica) [inline optuna/visualization/_parallel_coordinate.py:233]:
```suggestion
            vocab: defaultdict[int | str, int] = defaultdict(lambda: len(vocab))
```
It seems `defaultdict` is also available as a type annotation.
https://peps.python.org/pep-0585/#implementation

File: optuna/visualization/_parallel_coordinate.py [hunk 4]
```diff
@@ -1,15 +1,12 @@
+from __future__ import annotations
+
 from collections import defaultdict
 import math
 from typing import Any
 from typing import Callable
 from typing import cast
 from typing import DefaultDict
```
Comment (Alnusjaponica) [inline optuna/visualization/_parallel_coordinate.py:8]: Remove when `DefaultDict` is removed.
```suggestion
```

**Comment by Alnusjaponica** [comment]
Thank you for your contribution. I have found some annotations that could be further corrected, please let me know if I am wrong.

File: optuna/visualization/_contour.py (hunk 1 above)
Comment (harupy) [reply optuna/visualization/_contour.py:58]: It's just a matter of preference, but I think `X | None` is simpler than `Optional[X]`.

File: optuna/visualization/_parallel_coordinate.py (hunk 4 above)
Comment (harupy) [reply optuna/visualization/_parallel_coordinate.py]: Removed!

File: optuna/visualization/_parallel_coordinate.py (hunk 3 above)
Comment (harupy) [reply optuna/visualization/_parallel_coordinate.py]: Great suggestion. replaced.

File: optuna/visualization/_pareto_front.py (hunk 2 above)
Comment (harupy) [reply optuna/visualization/_pareto_front.py]: I think it's possible :)

**Review by Alnusjaponica — COMMENTED** [review]
Thank you for your update! The change itself LGTM. 
I am going to approve this PR after the cause of test failure (seems irrelevant to this PR) is identified.

**Comment by harupy** [comment]
The latest release of `distributed` might be related:
https://pypi.org/project/distributed/

**Comment by Alnusjaponica** [comment]
Thank you so much for the useful advice. I confirmed that the failure caused by the latest version of `distributed` in my local environment and created #4545 to temporary fix them.

**Comment by cross32768** [comment]
> Thank you so much for the useful advice. I confirmed that the failure caused by the latest version of `distributed` in my local environment and created #4545 to temporary fix them.

I merged #4545 and confirmed master branch is fixed. Could you merge master?

**Comment by harupy** [comment]
@cross32768 Rebased on master :)

**Pull request merged**